### PR TITLE
Remove variant propagation

### PR DESCRIPTION
**Pull request by mbj**
Depends on #321

### CODE REVIEW

File: api/spec/controllers/spree/api/orders_controller_spec.rb [hunk 1]
```diff
@@ -51,9 +51,9 @@ module Spree
           subject
           variant = json_response['line_items'][0]['variant']
           expect(variant).to_not be_nil
-          expect(variant['in_stock']).to eq(false)
-          expect(variant['total_on_hand']).to eq(0)
-          expect(variant['is_backorderable']).to eq(true)
+          expect(variant['in_stock']).to eq(true)
+          expect(variant['total_on_hand']).to eq(10)
+          expect(variant['is_backorderable']).to eq(false)
```
Comment (dkubb) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: You may want to use the `be(...)` matcher here and on the lines above.

Comment (mbj) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: Will do that, no GC happened yet and sorry for missing `in progress`.

File: core/app/models/spree/variant.rb [hunk 2]
```diff
@@ -242,13 +241,7 @@ def sellable?
...
         if product.master && product.master.in_stock?
           product.master.stock_items.update_all(:backorderable => false)
-          product.master.stock_items.each { |item| item.reduce_count_on_hand_to_zero }
-        end
-      end
-
-      def create_stock_items
-        StockLocation.where(propagate_all_variants: true).each do |stock_location|
-          stock_location.propagate_variant(self)
+          product.master.stock_items.each { |item| item.update_attributes!(count_on_hand: 0) }
```
Comment (dkubb) [inline core/app/models/spree/variant.rb]: What if we set the default `count_on_hand` to `0` in the database?

Comment (mbj) [inline core/app/models/spree/variant.rb]: Not this iteration. I removed `StockItem#reduce_count_on_hand_to_zero` which only had one call side (and terrible implicit specs) for this equivalent.

**Comment by mbj** [comment]
@dkubb This is ready for review after #321

File: api/spec/controllers/spree/api/orders_controller_spec.rb (hunk 1 above)
Comment (dkubb) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: @mbj ping

Comment (mbj) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: Meh, I missed these on GC.

File: core/app/models/spree/variant.rb (hunk 2 above)
Comment (dkubb) [inline core/app/models/spree/variant.rb]: Does this have to instantiate each stock item individually to update them, or can this use `#update_all` to mass change this across all stock items?

File: api/spec/controllers/spree/api/orders_controller_spec.rb (hunk 1 above)
Comment (mbj) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: mhh, weird, after checking out the branch and editing this file it shows correct. Maybe a GH caching issue!

File: core/lib/spree/testing_support/factories/variant_factory.rb [hunk 3]
```diff
@@ -17,10 +17,16 @@
...
 
     factory :variant do
-      # on_hand 5
       product { |p| p.association(:product) }
     end
 
+    factory :variant_in_stock do
+      after :create do |variant|
+        stock_location = Spree::StockLocation.first || create(:stock_location)
+        stock_location.stock_item_or_create(variant).update_attributes(count_on_hand: 10)
```
Comment (dkubb) [inline core/lib/spree/testing_support/factories/variant_factory.rb]: What about using `#update_attributes!` here?

Comment (mbj) [inline core/lib/spree/testing_support/factories/variant_factory.rb]: Jup, my fault missed.

File: backend/spec/features/admin/products/stock_management_spec.rb [hunk 4]
```diff
@@ -5,9 +5,8 @@
     visit(spree.admin_path)
   end
 
-  let!(:stock_location) { create(:stock_location, name: 'Default')                 }
-  let!(:product)        { create(:product, name: 'apache baseball cap', price: 10) }
-  let!(:variant)        { product.master                                           }
+  let!(:stock_location) { create(:stock_location)   }
```
Comment (dkubb) [inline backend/spec/features/admin/products/stock_management_spec.rb]: It looks like there's a few extra spaces before the `}` here.

File: core/spec/models/spree/order_inventory_spec.rb [hunk 5]
```diff
@@ -32,15 +32,14 @@
...
 
   context "#determine_target_shipment" do
-    let(:stock_location) { create :stock_location }
     let(:variant) { line_item.variant }
 
     before do
       subject.verify
 
-      order.shipments.create!(:stock_location_id => stock_location.id, :cost => 5)
+      order.shipments.create!(stock_location: Spree::StockLocation.first!, :cost => 5)
```
Comment (dkubb) [inline core/spec/models/spree/order_inventory_spec.rb]: You may want to use 1.9 style hashes here.

File: api/spec/controllers/spree/api/line_items_controller_spec.rb [hunk 6]
```diff
@@ -23,12 +23,12 @@ module Spree
 
       it "can update a line item on the order" do
         line_item = order.line_items.first!
-        api_put :update, :id => line_item.id, :line_item => { :quantity => 101 }
+        api_put :update, :id => line_item.id, :line_item => { :quantity => 10 }
```
Comment (dkubb) [inline api/spec/controllers/spree/api/line_items_controller_spec.rb]: You will want to use 1.9 style hashes here.

File: frontend/spec/features/checkout_spec.rb [hunk 7]
```diff
@@ -230,7 +230,7 @@
 
   # regression for #2921
   context "goes back from payment to add another item", js: true do
-    let!(:bag) { create(:product, :name => "RoR Bag") }
+    let!(:bag) { create(:product_in_stock, :name => "RoR Bag") }
```
Comment (dkubb) [inline frontend/spec/features/checkout_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/319/files#r63806727

File: frontend/spec/features/checkout_spec.rb [hunk 8]
```diff
@@ -290,7 +290,7 @@
     end
 
     context "and adds new product to cart and try to reach payment page" do
-      let!(:bag) { create(:product, :name => "RoR Bag") }
+      let!(:bag) { create(:product_in_stock, :name => "RoR Bag") }
```
Comment (dkubb) [inline frontend/spec/features/checkout_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/319/files#r63806727

**Comment by dkubb** [comment]
@mbj I finishes a pass over this PR and left a few notes.

**Comment by mbj** [comment]
@dkubb Ready for the next pass, all addressed.

File: core/app/models/spree/variant.rb (hunk 2 above)
Comment (mbj) [inline core/app/models/spree/variant.rb]: `#update_all` does not update the model instances, only the DB. At this point we should keep updating the model instances as it was in the original code.

Comment (mbj) [inline core/app/models/spree/variant.rb]: I'm looking for a clear 1:1 semantic refactoring on code scheduled for remove, not performing optimizations.

File: frontend/spec/support/shared_contexts/checkout_setup.rb [hunk 9]
```diff
@@ -3,7 +3,7 @@
   let!(:state) { create(:state, country: country) }
   let!(:shipping_method) { create(:shipping_method) }
   let!(:stock_location) { create(:stock_location) }
-  let!(:mug) { create(:product, name: "RoR Mug") }
+  let!(:mug) { create(:product_in_stock, name: "RoR Mug") }
```
Comment (dkubb) [inline frontend/spec/support/shared_contexts/checkout_setup.rb]: On line https://github.com/MountainRoseHerbs/spree/pull/319/files#diff-ef75e4c04d47a7559e3e2b6e107e9ec4R233 you changed the double quotes to single quotes. I would suggest doing the same here for consistency.

**Pull request merged**